### ISSUE TITLE
Use Zstd instead of gzip for compression

### ISSUE THREAD

**Issue by jrmuizel** [mozilla/sccache#552]
We're currently spending an appreciable amount of cpu time doing zlib compression:
```
  32.67%  sccache-dist     sccache-dist        [.] miniz_oxide::deflate::core::compress_inner
  11.67%  sccache-dist     sccache-dist        [.] miniz_oxide::deflate::core::compress_block
```
Using Zstd should give better compression and better performance.

**Comment by jrmuizel** [comment]
FWIW, icecream uses lzo which is also a lot faster.

**Comment by froydnj** [comment]
Isn't that just because `miniz_oxide` is written in pure Rust without the massive optimization work that's gone into zlib?

**Comment by jrmuizel** [comment]
I'm not sure how the performance of miniz_oxide compares to zlib, but even zlib is appreciably slower than Zstd.

**Comment by luser** [comment]
Currently cache entries are just zip files. It wouldn't be hard to change that, all the code for creating and unpacking cache entries is in this file:
https://github.com/mozilla/sccache/blob/d3093392af6f593361c229deb36dd14c964c0778/src/cache/cache.rs

You'd need a container format of some sort since there can be multiple output files + stdout/stderr, but it should be straightforward to use tar files.

**Comment by luser** [comment]
Oh, it strikes me that since you're using `sccache-dist` there this is probably toolchain/inputs/outputs packaging.

Toolchain:
https://github.com/mozilla/sccache/blob/d3093392af6f593361c229deb36dd14c964c0778/src/dist/pkg.rs#L146-L159

Inputs packaging:
https://github.com/mozilla/sccache/blob/0cb2e77cd005ea027fcd102ade6b6158bbc765b0/src/dist/http.rs#L1246
Inputs unpackaging:
https://github.com/mozilla/sccache/blob/d3093392af6f593361c229deb36dd14c964c0778/src/dist/http.rs#L980

Outputs are handled here:
https://github.com/mozilla/sccache/blob/d3093392af6f593361c229deb36dd14c964c0778/src/dist/mod.rs#L424-L446

Unfortunately if you change this you'll either need to support both compression formats for interop or change the whole dist API to a v2.

**Comment by luser** [comment]
Related, while looking at that code I realized that at least for Rust compilation packaging compile inputs is probably doing a lot of redundant work so I filed this issue on a potential optimization: https://github.com/mozilla/sccache/issues/558

**Comment by froydnj** [comment]
A clobber build that solely distributes compiles to a remote does appear to spend a significant amount of time in the local `sccache` server writing cache entries and therefore doing a bunch of deflate compression (50%, I think?).  `perf` is being its usual unhelpful self, but I do not see any `flate2`-related stuff regarding any of the dist API in the performance profile.

I guess we could modify [zip-rs](https://github.com/mvdnes/zip-rs) to support zstd for individual data entries in the zip file?  It's unmaintained, which I guess would be a good thing for something like this.  I don't know that I'd feel comfortable hard-coding some constant in zip entries to be zstd, so we'd have to figure out how to plumb that down into the guts of `ZipArchive` with some kind of trait?  (Something like [rc-zip](https://crates.io/crates/rc-zip) which is only a "statement of intent" at the momemt.)

**Comment by froydnj** [comment]
> A clobber build that solely distributes compiles to a remote does appear to spend a significant amount of time in the local `sccache` server writing cache entries and therefore doing a bunch of deflate compression (50%, I think?).

I should note that these measurements *were* taken on a binary that changed input compression to be done with zstd...but I'll also note that changing `zip` and `flate2` to use native zlib (while still using zstd compression for the inputs) made deflate compression basically disappear from the profile (!).  Maybe that's just `perf` being stupid (it insists on telling me cache entry compression and writing is being done from the kernel's interrupt handler...), but maybe not.

I think I'll try putting together a patch that enables native zlib at the very least; I think that should go in regardless.  We'll see how things look and whether zstd compression should be enabled for the inputs.

**Comment by FooBarWidget** [comment]
I think this issue can be closed now that #784 is merged.

**Comment by sylvestre** [comment]
Thanks
